### ISSUE TITLE
Degeneracy question

### ISSUE THREAD

**Issue by afonari** [msg-byu/enumlib#61]
I have a structure with 16 atoms and partial concentration:
```
Title
bulk
5.56917 0.00000 0.00000
2.78459 4.82305 0.00000
-0.00000 -0.00000 27.28326
2
16
6.96147 4.01920 4.54721 0/1
4.17688 4.01920 4.54721 0/1
5.56917 1.60768 4.54721 0/1
2.78459 1.60768 4.54721 0/1
-0.00000 -0.00000 6.82082 0/1
1.39229 0.80384 9.09442 0/1
4.17688 0.80384 9.09442 0/1
2.78459 3.21536 9.09442 0/1
5.56917 3.21536 9.09442 0/1
6.96147 4.01920 11.36803 0/1
4.17688 4.01920 11.36803 0/1
5.56917 1.60768 11.36803 0/1
2.78459 1.60768 11.36803 0/1
4.17688 2.41152 6.82082 0/1
1.39229 2.41152 6.82082 0/1
2.78459 -0.00000 6.82082 0/1
1 1
0.001
partial
12 12 16
4 4 16
```

I get struct_enum.out as follows:
```
 # <fortpy version="3" revision="247"></fortpy>
Title
bulk
  5.5691700       0.0000000       0.0000000        # a1 parent lattice vector
  2.7845900       4.8230500       0.0000000        # a2 parent lattice vector
  0.0000000       0.0000000       27.283260        # a3 parent lattice vector
   16 # Number of points in the multilattice
  6.9614700       4.0192000       4.5472100        # d01 d-vector, labels: 0/1
  4.1768800       4.0192000       4.5472100        # d02 d-vector, labels: 0/1
  5.5691700       1.6076800       4.5472100        # d03 d-vector, labels: 0/1
  2.7845900       1.6076800       4.5472100        # d04 d-vector, labels: 0/1
  0.0000000       0.0000000       6.8208200        # d05 d-vector, labels: 0/1
  1.3922900      0.80384000       9.0944200        # d06 d-vector, labels: 0/1
  4.1768800      0.80384000       9.0944200        # d07 d-vector, labels: 0/1
  2.7845900       3.2153600       9.0944200        # d08 d-vector, labels: 0/1
  5.5691700       3.2153600       9.0944200        # d09 d-vector, labels: 0/1
  6.9614700       4.0192000       11.368030        # d10 d-vector, labels: 0/1
  4.1768800       4.0192000       11.368030        # d11 d-vector, labels: 0/1
  5.5691700       1.6076800       11.368030        # d12 d-vector, labels: 0/1
  2.7845900       1.6076800       11.368030        # d13 d-vector, labels: 0/1
  4.1768800       2.4115200       6.8208200        # d14 d-vector, labels: 0/1
  1.3922900       2.4115200       6.8208200        # d15 d-vector, labels: 0/1
  2.7845900       0.0000000       6.8208200        # d16 d-vector, labels: 0/1
 2-nary case
   1   1 # Starting and ending cell sizes for search
 0.10000000E-02 # Epsilon (finite precision parameter)
Concentration check:
    T
Including only structures of which the concentration   of each atom is in the range:
Type 1:   12/  16 --   12/  16
Type 2:    4/  16 --    4/  16
partial list of labelings (complete labelings only) is used
Equivalency list: 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16
start   #tot      HNF     Hdegn   labdegn   Totdegn   #size idx    pg    SNF             HNF                 Left transform                          labeling
          1         1       1        0        0           1   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1111000000000000
          2         1       1        0        0           2   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1110100000000000
          3         1       1        0        0           3   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1110010000000000
          4         1       1        0        0           4   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1110000010000000
          5         1       1        0        0           5   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1110000001000000
          6         1       1        0        0           6   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1110000000001000
          7         1       1        0        0           7   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1110000000000100
          8         1       1        0        0           8   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100110000000000
          9         1       1        0        0           9   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100100100000000
         10         1       1        0        0          10   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100100010000000
         11         1       1        0        0          11   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100100001000000
         12         1       1        0        0          12   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100100000010000
         13         1       1        0        0          13   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100100000001000
         14         1       1        0        0          14   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100100000000100
         15         1       1        0        0          15   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100100000000001
         16         1       1        0        0          16   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100011000000000
         17         1       1        0        0          17   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100010100000000
         18         1       1        0        0          18   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100010001000000
         19         1       1        0        0          19   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100010000100000
         20         1       1        0        0          20   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100010000010000
         21         1       1        0        0          21   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100010000000100
         22         1       1        0        0          22   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100010000000010
         23         1       1        0        0          23   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000110000000
         24         1       1        0        0          24   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000101000000
         25         1       1        0        0          25   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000100010000
         26         1       1        0        0          26   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000100001000
         27         1       1        0        0          27   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000100000100
         28         1       1        0        0          28   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000001100000
         29         1       1        0        0          29   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000001010000
         30         1       1        0        0          30   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000001000100
         31         1       1        0        0          31   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000001000010
         32         1       1        0        0          32   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000000011000
         33         1       1        0        0          33   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000000010100
         34         1       1        0        0          34   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1100000000000110
         35         1       1        0        0          35   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000111000000000
         36         1       1        0        0          36   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000110010000000
         37         1       1        0        0          37   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000110001000000
         38         1       1        0        0          38   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000110000100000
         39         1       1        0        0          39   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000110000001000
         40         1       1        0        0          40   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000110000000100
         41         1       1        0        0          41   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000110000000010
         42         1       1        0        0          42   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101100000000
         43         1       1        0        0          43   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101010000000
         44         1       1        0        0          44   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101001000000
         45         1       1        0        0          45   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101000100000
         46         1       1        0        0          46   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101000010000
         47         1       1        0        0          47   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101000000100
         48         1       1        0        0          48   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101000000010
         49         1       1        0        0          49   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000101000000001
         50         1       1        0        0          50   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100011000000
         51         1       1        0        0          51   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100010100000
         52         1       1        0        0          52   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100010001000
         53         1       1        0        0          53   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100010000100
         54         1       1        0        0          54   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100001000100
         55         1       1        0        0          55   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100001000010
         56         1       1        0        0          56   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100000100100
         57         1       1        0        0          57   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100000100010
         58         1       1        0        0          58   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100000100001
         59         1       1        0        0          59   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100000001100
         60         1       1        0        0          60   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100000000110
         61         1       1        0        0          61   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000100000000011
         62         1       1        0        0          62   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000011100000000
         63         1       1        0        0          63   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000011000000100
         64         1       1        0        0          64   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000010001000100
         65         1       1        0        0          65   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000001110000000
         66         1       1        0        0          66   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000001100000100
         67         1       1        0        0          67   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   1000001000100100
         68         1       1        0        0          68   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   0000111100000000
         69         1       1        0        0          69   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   0000111010000000
         70         1       1        0        0          70   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   0000111000000100
         71         1       1        0        0          71   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   0000111000000001
         72         1       1        0        0          72   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   0000110010000100
         73         1       1        0        0          73   1    48   1  1  1    1  0  1  0  0  1      1    0    0    0    1    0    0    0    1   0000100000000111
```

73 structures with 0 degeneracy each. Based on the formula N=16!/(4!*12!)=1820 the total number of combinations should be 1820.

Can anyone please help me understand this, thanks!

**Comment by wsmorgan** [comment]
@afonari Enumlib has two different algorithms for determining the unique structures, one of them constructs a list of all possible configurations and crosses off duplicates as it finds them enabling the code to count and report the number of degenerate structures, the second uses a tree search to eliminate configurations without ever knowing what the full list of configurations is (see paper http://msg.byu.edu/papers/recStabEnumeration.pdf for details), due to the nature of the second algorithm it is not possible to know how many structures are degenerate to any given structure. The second algorithm is faster and so whenever possible it is the algorithm the code will use. When the second algorithm is used the degeneracies in the output file are set to zero since they are unknown. Do you need the degeneracies?

**Comment by afonari** [comment]
Hi @wsmorgan, thanks for the fast reply. I think in this case, first (old algorithm) algorithm is used (due to this condition https://github.com/msg-byu/enumlib/blob/9fdf5476d66a0adf5e61ae9ecefdac3cbaa11b6e/src/derivative_structure_generator.f90#L1486 (see also your comment on Jan 18 in https://github.com/msg-byu/enumlib/issues/43)

**Comment by wsmorgan** [comment]
@afonari The difference between your use case and the usage brought up there is site restrictions, you're allowing atom type A and atom type B to occupy any site on the lattice while in that case some of the sites were extremely restricted. In the code:
```
                   if (any(site_res == 0) .and. (multinomial(iRange(iC,:)) < max_binomial) .and. (arrows .eqv. .false.)) then
```
the statement `any(site_res == 0)` only evaluates to true if site restrictions were given. So in your case even if the multinomial is large you'll end up in the new algorithm. The old one only gets used for site restrictions.

**Comment by afonari** [comment]
Ah, yes, you are correct, just confirmed it by printing the algorithm name in the source code.

Back to my question, for the new algorithm, is there a way to determine degeneracy?

EDIT: just confirmed that if old algorithm is requested (by changing the source code), degeneracies are correct. Their sum is 1820 as expected.

**Comment by wsmorgan** [comment]
No way that wouldn't drastically increase the computational cost (we considered doing this once but the only way we devised at the time was to create a list of all possible configurations and then when the algorithm found a unique one it would cross off duplicates getting the degeneracy, but this is exactly the old algorithm so we didn't take that approach). Do you need the degeneracy in your work?

**Comment by afonari** [comment]
> Do you need the degeneracy in your work?

Yes, degeneracy can be used as a meaning of weight in ensemble averages, see for example: http://ecst.ecsdl.org/content/25/1/1335.abstract

**Comment by wsmorgan** [comment]
Then you'll need to force the code into using the old algorithm.

@glwhart  Maybe we should consider an input flag that would force this behavior, or perhaps we should revisit the degeneracy problem and see if we can determine an elegant solution?

**Comment by afonari** [comment]
Thanks @wsmorgan for explanation and possible solution. Will take a look into "consider an input flag that would force this behavior".

**Comment by glwhart** [comment]
There might be a way to take a set of structures and *compute* the
degeneracy. That would be far more efficient...assuming the set of
structures was small...

Gus Hart
http://msg.byu.edu
Assoc. Dean, CPMS
http://cpms.byu.edu

**Comment by glwhart** [comment]
Closing this as it has been merged with #62
